File: server/rest/src/main/scala/org/infinispan/rest/MimeMetadata.scala
package org.infinispan.rest

import org.infinispan.container.versioning.EntryVersion
import java.util.concurrent.TimeUnit.{MILLISECONDS => MILLIS}
import java.util.concurrent.TimeUnit
import org.infinispan.metadata.{EmbeddedMetadata, Metadata}
import Metadata.Builder
import org.infinispan.commons.marshall.AbstractExternalizer
import java.util
import java.io.{ObjectInput, ObjectOutput}
import scala.collection.JavaConversions.setAsJavaSet
import org.jboss.marshalling.util.IdentityIntMap

/**
 * Metadata for MIME data stored in REST servers.
 *
 * @author Galder Zamarreño
 * @since 5.3
 */
class MimeMetadata(val contentType: String) extends Metadata {

   def lifespan(): Long = -1

   def maxIdle(): Long = -1

   def version(): EntryVersion = null

   def builder(): Builder = new MimeMetadataBuilder().contentType(contentType)

   override def equals(obj: Any): Boolean = {
      obj match {
         case that: MimeMetadata =>
            that.canEqual(this) && contentType == that.contentType
         case _ => false
      }
   }

   def canEqual(other: Any): Boolean = other.isInstanceOf[MimeMetadata]

   override def hashCode(): Int = 41 + contentType.hashCode

   override def toString: String = s"MimeMetadata(contentType=$contentType)"

}

private class MimeExpirableMetadata(override val contentType: String,
        lifespanTime: Long, lifespanUnit: TimeUnit,
        maxIdleTime: Long, maxIdleUnit: TimeUnit) extends MimeMetadata(contentType) {

   override final val lifespan = lifespanUnit.toMillis(lifespanTime)

   override final val maxIdle = maxIdleUnit.toMillis(maxIdleTime)

   override def builder(): Builder = new MimeMetadataBuilder()
         .contentType(contentType).lifespan(lifespan).maxIdle(maxIdle)

   override def equals(obj: Any): Boolean = {
      obj match {
         case that: MimeExpirableMetadata =>
            that.canEqual(this) &&
                    contentType == that.contentType &&
                    lifespan == that.lifespan &&
                    maxIdle == that.maxIdle
         case _ => false
      }
   }

   override def canEqual(other: Any): Boolean = other.isInstanceOf[MimeExpirableMetadata]

   override def hashCode(): Int =
      41 * (41 * (41 + contentType.hashCode) + lifespan.toInt) + maxIdle.toInt

   override def toString: String =
      s"MimeExpirableMetadata(contentType=$contentType, lifespan=$lifespan, maxIdle=$maxIdle)"

}

private class MimeLifespanExpirableMetadata(override val contentType: String,
      lifespanTime: Long, lifespanUnit: TimeUnit) extends MimeMetadata(contentType) {

   override final val lifespan = lifespanUnit.toMillis(lifespanTime)

   override def builder(): Builder = new MimeMetadataBuilder()
         .contentType(contentType).lifespan(lifespan)

   override def equals(obj: Any): Boolean = {
      obj match {
         case that: MimeExpirableMetadata =>
            that.canEqual(this) &&
                  contentType == that.contentType &&
                  lifespan == that.lifespan
         case _ => false
      }
   }

   override def canEqual(other: Any): Boolean = other.isInstanceOf[MimeLifespanExpirableMetadata]

   override def hashCode(): Int =
      41 * (41 + contentType.hashCode) + lifespan.toInt

   override def toString: String =
      s"MimeLifespanExpirableMetadata(contentType=$contentType, lifespan=$lifespan)"

}

private class MimeMaxIdleExpirableMetadata(override val contentType: String,
      maxIdleTime: Long, maxIdleUnit: TimeUnit) extends MimeMetadata(contentType) {

   override final val maxIdle = maxIdleUnit.toMillis(maxIdleTime)

   override def builder(): Builder = new MimeMetadataBuilder()
         .contentType(contentType).maxIdle(maxIdle)

   override def equals(obj: Any): Boolean = {
      obj match {
         case that: MimeExpirableMetadata =>
            that.canEqual(this) &&
                  contentType == that.contentType &&
                  maxIdle == that.maxIdle
         case _ => false
      }
   }

   override def canEqual(other: Any): Boolean = other.isInstanceOf[MimeMaxIdleExpirableMetadata]

   override def hashCode(): Int =
      41 * (41 + contentType.hashCode) + maxIdle.toInt

   override def toString: String =
      s"MimeMaxIdleExpirableMetadata(contentType=$contentType, maxIdle=$maxIdle)"

}

class MimeMetadataBuilder extends EmbeddedMetadata.Builder {

   private var contentType: String = _

   def contentType(contentType: String): MimeMetadataBuilder = {
      this.contentType = contentType
      this
   }

   override def build(): Metadata = {
      new MimeExpirableMetadata(contentType, lifespan, lifespanUnit, maxIdle, maxIdleUnit)
   }
}

object MimeMetadata {

   private def apply(contentType: String): MimeMetadata = new MimeMetadata(contentType)

   class Externalizer extends AbstractExternalizer[MimeMetadata] {

      final val Immortal = 0
      final val Expirable = 1
      final val LifespanExpirable = 2
      final val MaxIdleExpirable = 3

      final val numbers = new IdentityIntMap[Class[_]](4)

      numbers.put(classOf[MimeMetadata], Immortal)
      numbers.put(classOf[MimeExpirableMetadata], Expirable)
      numbers.put(classOf[MimeLifespanExpirableMetadata], LifespanExpirable)
      numbers.put(classOf[MimeMaxIdleExpirableMetadata], MaxIdleExpirable)

      def readObject(input: ObjectInput): MimeMetadata = {
         val contentType = input.readUTF()
         val number = input.readUnsignedByte()
         number match {
            case Immortal => MimeMetadata(contentType)
            case Expirable =>
               val lifespan = input.readLong()
               val maxIdle = input.readLong()
               new MimeExpirableMetadata(contentType, lifespan, MILLIS, maxIdle, MILLIS)
            case LifespanExpirable =>
               val lifespan = input.readLong()
               new MimeLifespanExpirableMetadata(contentType, lifespan, MILLIS)
            case MaxIdleExpirable =>
               val maxIdle = input.readLong()
               new MimeMaxIdleExpirableMetadata(contentType, maxIdle, MILLIS)
         }
      }

      def writeObject(output: ObjectOutput, meta: MimeMetadata) {
         output.writeUTF(meta.contentType)
         val number = numbers.get(meta.getClass, -1)
         output.write(number)
         number match {
            case Immortal => // no-op
            case Expirable =>
               output.writeLong(meta.lifespan())
               output.writeLong(meta.maxIdle())
            case LifespanExpirable =>
               output.writeLong(meta.lifespan())
            case MaxIdleExpirable =>
               output.writeLong(meta.maxIdle())
         }
      }

      def getTypeClasses: util.Set[Class[_ <: MimeMetadata]] =
         setAsJavaSet(Set[java.lang.Class[_ <: MimeMetadata]](
            classOf[MimeMetadata], classOf[MimeExpirableMetadata],
            classOf[MimeLifespanExpirableMetadata], classOf[MimeMaxIdleExpirableMetadata]))

   }

}

File: server/rest/src/main/java/org/infinispan/rest/operations/mime/MimeMetadata.java
package org.infinispan.rest.operations.mime;

import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectOutput;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import org.infinispan.commands.CommandInvocationId;
import org.infinispan.commands.InvocationRecord;
import org.infinispan.commons.marshall.AbstractExternalizer;
import org.infinispan.commons.util.Util;
import org.infinispan.container.versioning.EntryVersion;
import org.infinispan.metadata.Metadata;
import org.jboss.marshalling.util.IdentityIntMap;

/**
 * Metadata for MIME data stored in REST servers.
 *
 * @author Galder Zamarreño
 * @since 5.3
 */
public class MimeMetadata implements Metadata {
   protected final String contentType;
   protected final InvocationRecord records;

   public MimeMetadata(String contentType, InvocationRecord records) {
      this.contentType = Objects.requireNonNull(contentType);
      this.records = records;
   }

   @Override
   public long lifespan() {
      return -1;
   }

   @Override
   public long maxIdle() {
      return -1;
   }

   @Override
   public EntryVersion version() {
      return null;
   }

   @Override
   public InvocationRecord lastInvocation() {
      return records;
   }

   @Override
   public InvocationRecord invocation(CommandInvocationId id) {
      return InvocationRecord.lookup(records, id);
   }

   @Override
   public Builder builder() {
      return new MimeMetadataBuilder().contentType(contentType).invocations(records);
   }

   public String contentType() {
      return contentType;
   }

   @Override
   public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;

      MimeMetadata that = (MimeMetadata) o;

      return contentType.equals(that.contentType);

   }

   @Override
   public int hashCode() {
      return 41 + contentType.hashCode();
   }

   @Override
   public String toString() {
      return "MimeMetadata{" +
            "contentType='" + contentType + '\'' +
            '}';
   }

   private static final int Immortal = 0;
   private static final int Expirable = 1;
   private static final int LifespanExpirable = 2;
   private static final int MaxIdleExpirable = 3;

   private static final IdentityIntMap<Class> numbers = new IdentityIntMap<>(4);

   static {
      numbers.put(MimeMetadata.class, Immortal);
      numbers.put(MimeExpirableMetadata.class, Expirable);
      numbers.put(MimeLifespanExpirableMetadata.class, LifespanExpirable);
      numbers.put(MimeMaxIdleExpirableMetadata.class, MaxIdleExpirable);
   }

   public static class Externalizer extends AbstractExternalizer<MimeMetadata> {

      @Override
      public MimeMetadata readObject(ObjectInput input) throws IOException, ClassNotFoundException {
         String contentType = input.readUTF();
         InvocationRecord records = InvocationRecord.readListFrom(input);
         int number = input.readUnsignedByte();
         switch (number) {
            case Immortal:
               return new MimeMetadata(contentType, records);
            case Expirable:
               long lifespan = input.readLong();
               long maxIdle = input.readLong();
               return new MimeExpirableMetadata(contentType, records, lifespan, TimeUnit.MILLISECONDS, maxIdle, TimeUnit.MILLISECONDS);
            case LifespanExpirable:
               lifespan = input.readLong();
               return new MimeLifespanExpirableMetadata(contentType, records, lifespan, TimeUnit.MILLISECONDS);
            case MaxIdleExpirable:
               maxIdle = input.readLong();
               return new MimeMaxIdleExpirableMetadata(contentType, records, maxIdle, TimeUnit.MILLISECONDS);
            default:
               throw new IllegalArgumentException("Unsupported type: " + number);
         }
      }

      @Override
      public void writeObject(ObjectOutput output, MimeMetadata meta) throws IOException {
         output.writeUTF(meta.contentType);
         InvocationRecord.writeListTo(output, meta.records);
         int number = numbers.get(meta.getClass(), -1);
         output.write(number);
         switch (number) {
            case Immortal:
               // no-op
               break;
            case Expirable:
               output.writeLong(meta.lifespan());
               output.writeLong(meta.maxIdle());
               break;
            case LifespanExpirable:
               output.writeLong(meta.lifespan());
               break;
            case MaxIdleExpirable:
               output.writeLong(meta.maxIdle());
               break;
            default:
               throw new IllegalArgumentException("Unsupported type: " + number);
         }
      }

      @Override
      public Set<Class<? extends MimeMetadata>> getTypeClasses() {
         return Util.asSet(MimeMetadata.class, MimeExpirableMetadata.class,
               MimeLifespanExpirableMetadata.class, MimeMaxIdleExpirableMetadata.class);
      }
   }
}

class MimeExpirableMetadata extends MimeMetadata {
   private final long lifespanTime;
   private final TimeUnit lifespanUnit;
   private final long maxIdleTime;
   private final TimeUnit maxIdleUnit;

   public MimeExpirableMetadata(String contentType, InvocationRecord records, long lifespanTime, TimeUnit lifespanUnit, long maxIdleTime,
                                TimeUnit maxIdleUnit) {
      super(contentType, records);
      this.lifespanTime = lifespanTime;
      this.lifespanUnit = Objects.requireNonNull(lifespanUnit);
      this.maxIdleTime = maxIdleTime;
      this.maxIdleUnit = Objects.requireNonNull(maxIdleUnit);
   }

   @Override
   public long lifespan() {
      return lifespanUnit.toMillis(lifespanTime);
   }

   @Override
   public long maxIdle() {
      return maxIdleUnit.toMillis(maxIdleTime);
   }

   @Override
   public Builder builder() {
      return super.builder().lifespan(lifespanTime, lifespanUnit).maxIdle(maxIdleTime, maxIdleUnit);
   }

   @Override
   public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      if (!super.equals(o)) return false;

      MimeExpirableMetadata that = (MimeExpirableMetadata) o;

      if (lifespanTime != that.lifespanTime) return false;
      if (maxIdleTime != that.maxIdleTime) return false;
      if (lifespanUnit != that.lifespanUnit) return false;
      return maxIdleUnit == that.maxIdleUnit;

   }

   @Override
   public int hashCode() {
      int result = super.hashCode();
      result = 31 * result + (int) (lifespanTime ^ (lifespanTime >>> 32));
      result = 31 * result + lifespanUnit.hashCode();
      result = 31 * result + (int) (maxIdleTime ^ (maxIdleTime >>> 32));
      result = 31 * result + maxIdleUnit.hashCode();
      return result;
   }

   @Override
   public String toString() {
      return "MimeExpirableMetadata{" +
            "lifespanTime=" + lifespanTime +
            ", lifespanUnit=" + lifespanUnit +
            ", maxIdleTime=" + maxIdleTime +
            ", maxIdleUnit=" + maxIdleUnit +
            '}';
   }
}

class MimeLifespanExpirableMetadata extends MimeMetadata {
   private final long lifespanTime;
   private final TimeUnit lifespanUnit;

   MimeLifespanExpirableMetadata(String contentType, InvocationRecord records, long lifespanTime, TimeUnit lifespanUnit) {
      super(contentType, records);
      this.lifespanTime = lifespanTime;
      this.lifespanUnit = Objects.requireNonNull(lifespanUnit);
   }

   @Override
   public long lifespan() {
      return lifespanUnit.toMillis(lifespanTime);
   }

   @Override
   public Builder builder() {
      return super.builder().lifespan(lifespanTime, lifespanUnit);
   }

   @Override
   public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      if (!super.equals(o)) return false;

      MimeLifespanExpirableMetadata that = (MimeLifespanExpirableMetadata) o;

      if (lifespanTime != that.lifespanTime) return false;
      return lifespanUnit == that.lifespanUnit;

   }

   @Override
   public int hashCode() {
      int result = super.hashCode();
      result = 31 * result + (int) (lifespanTime ^ (lifespanTime >>> 32));
      result = 31 * result + lifespanUnit.hashCode();
      return result;
   }

   @Override
   public String toString() {
      return "MimeLifespanExpirableMetadata{" +
            "lifespanTime=" + lifespanTime +
            ", lifespanUnit=" + lifespanUnit +
            '}';
   }
}

class MimeMaxIdleExpirableMetadata extends MimeMetadata {
   private final long maxIdleTime;
   private final TimeUnit maxIdleUnit;

   public MimeMaxIdleExpirableMetadata(String contentType, InvocationRecord records, long maxIdleTime, TimeUnit maxIdleUnit) {
      super(contentType, records);
      this.maxIdleTime = maxIdleTime;
      this.maxIdleUnit = Objects.requireNonNull(maxIdleUnit);
   }

   @Override
   public long maxIdle() {
      return maxIdleUnit.toMillis(maxIdleTime);
   }

   @Override
   public Builder builder() {
      return super.builder().maxIdle(maxIdleTime, maxIdleUnit);
   }

   @Override
   public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      if (!super.equals(o)) return false;

      MimeMaxIdleExpirableMetadata that = (MimeMaxIdleExpirableMetadata) o;

      if (maxIdleTime != that.maxIdleTime) return false;
      return maxIdleUnit == that.maxIdleUnit;

   }

   @Override
   public int hashCode() {
      int result = super.hashCode();
      result = 31 * result + (int) (maxIdleTime ^ (maxIdleTime >>> 32));
      result = 31 * result + maxIdleUnit.hashCode();
      return result;
   }

   @Override
   public String toString() {
      return "MimeMaxIdleExpirableMetadata{" +
            "maxIdleTime=" + maxIdleTime +
            ", maxIdleUnit=" + maxIdleUnit +
            '}';
   }
}
